src/controllers/ProductsController.php

```diff
@@ -40,14 +40,30 @@
  * @author Pixel & Tonic, Inc. <support@pixelandtonic.com>
  * @since 2.0
  */
-class ProductsController extends BaseCpController
+class ProductsController extends BaseController
 {
+    /**
+     * @var string[] The action names that bypass the "Access Craft Commerce" permission.
+     */
+    protected $ignorePluginPermission = ['save-product', 'duplicate-product', 'delete-product'];
+
     /**
      * @inheritdoc
      */
     public function init()
     {
         parent::init();
+
+        $actionSegs = Craft::$app->getRequest()->getActionSegments();
+
+        $actionName = $actionSegs[2] ?? null;
+
+        $requireAccessPluginCommerce = ($actionName !== null && in_array($actionName, $this->ignorePluginPermission)) ? false : true;
+
+        if ($requireAccessPluginCommerce === true) {
+            $this->requirePermission('accessPlugin-commerce');
+        }
+
```

Comment:
Just realised it might be easier to use the beforeAction method which gives you the action being run directly. Example:

https://github.com/craftcms/cms/blob/develop/src/controllers/UsersController.php#L118-L127

So something like below (untested):

```php
    /**
     * @inheritdoc
     */
    public function beforeAction($action)
    {
        if (!parent::beforeAction($action)) {
            return false;
        }

        if (!in_array($action->id, ['save-product', 'delete-product', 'duplicate-product'])) {
            $this->requirePermission('accessPlugin-commerce');
        }

        if (in_array($action->id, ['save-product', 'delete-product', 'duplicate-product'])) {
            $this->requirePermission('commerce-manageProducts');
        }

        return true;
    }
```

You can then remove the `init()`

Reply:
Another example in cms repo: https://github.com/craftcms/cms/blob/develop/src/controllers/ProjectConfigController.php#L31-L39

Reply:
I think we don't need to put ```$this->requirePermission('commerce-manageProducts');``` in a condition because we want to have that permission on all the methods